src/solana-wallet-adapter/adapter.ts

```diff
@@ -0,0 +1,128 @@
+import {
+    BaseMessageSignerWalletAdapter,
+    WalletNotConnectedError,
+    WalletNotReadyError,
+    WalletPublicKeyError,
+    WalletReadyState,
+    WalletSignTransactionError,
+    WalletWindowClosedError,
+} from "@solana/wallet-adapter-base";
+import { PublicKey, Transaction } from "@solana/web3.js";
+
+import CrossmintEmbed from "../CrossmintEmbed";
+import { CROSSMINT_LOGO_21x21, CrossmintWalletName } from "../consts/branding";
+import { BlockchainTypes, CrossmintEmbedConfig, CrossmintEmbedParams } from "../types";
+import { buildConfig } from "../utils/config";
+
+export class CrossmintSolanaWalletAdapter extends BaseMessageSignerWalletAdapter {
+    name = CrossmintWalletName;
+    url = "https://www.crossmint.io";
+    icon = CROSSMINT_LOGO_21x21;
+
+    private _connecting: boolean;
+    private _publicKey: PublicKey | null;
+    private _readyState: WalletReadyState =
+        typeof window === "undefined" ? WalletReadyState.Unsupported : WalletReadyState.Loadable;
+
+    private _config: CrossmintEmbedConfig;
+    private _client?: CrossmintEmbed;
+
+    constructor(params: Omit<CrossmintEmbedParams, "chain">) {
+        super();
+
+        this._connecting = false;
+        this._publicKey = null;
+
+        this._config = buildConfig({ ...params, chain: BlockchainTypes.SOLANA });
+    }
+
+    get publicKey(): PublicKey | null {
+        return this._publicKey;
+    }
+
+    get connecting(): boolean {
+        return this._connecting;
+    }
+
+    get connected(): boolean {
+        return this._publicKey !== null && this._publicKey !== undefined;
+    }
+
+    get readyState(): WalletReadyState {
+        return this._readyState;
+    }
+
+    async connect(): Promise<void> {
+        try {
+            if (this.connected || this.connecting) return;
+            if (this._readyState !== WalletReadyState.Loadable) throw new WalletNotReadyError();
+
+            this._connecting = true;
+
+            const client = await CrossmintEmbed.init(this._config);
+
+            const account = await client.login();
+
+            if (account === null) {
+                throw new WalletWindowClosedError("User rejected the request");
+            }
+            if (account === undefined) {
+                throw new WalletNotConnectedError("User rejected the request or closed the window");
+            }
```

Comment:
Woah, if this is why we use null vs undefined, maybe we can throw custom exceptions instead? Even just these exceptions directly?

Reply:
I didnt want to do that so we can can keep the architecture agnostic (these error types are specific for @solana/wallet-adapter). We could make our own set of errors though